### PR TITLE
feat(component_interface_utils): add interface classes 

### PR DESCRIPTION
**Pull request by isamu-takagi**
## Description

* Add wrapper classes to the component_interface_utils package. AD API uses these classes later.
* Apply this change to interface version.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: system/default_ad_api/src/interface.cpp [hunk 1]
```diff
@@ -12,27 +12,26 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "default_ad_api/nodes/interface_version.hpp"
+#include "interface.hpp"
```
Comment (kenji-miyake) [inline system/default_ad_api/src/interface.cpp:15]: I believe you should split this refactoring into another PR for reviewers.

File: common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#define COMPONENT_INTERFACE_UTILS__RCLCPP__SERVICE_CLIENT_HPP_
+
+#include <rclcpp/client.hpp>
+
+#include <utility>
+
+namespace component_interface_utils
+{
+
+/// The wrapper class of rclcpp::Client for logging.
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:27]: I forgot it. Added https://github.com/autowarefoundation/autoware.universe/pull/899/commits/13538d272b5dc009893c379361eeb36a6974a2fd

File: common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+template <class SpecT>
+class Client
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:29]: To prevent unintentional use of superclass members, such as not outputting logs.

File: common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+#ifndef COMPONENT_INTERFACE_UTILS__RCLCPP__TOPIC_PUBLISHER_HPP_
+#define COMPONENT_INTERFACE_UTILS__RCLCPP__TOPIC_PUBLISHER_HPP_
+
+#include <rclcpp/publisher.hpp>
+
+namespace component_interface_utils
+{
+
+/// The wrapper class of rclcpp::Publisher.
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp:23]: Now, this only holds type information by `using SpecType`.
Mainly for the future use (e.g. publish only when data changing, topic rate check).

File: common/component_interface_utils/include/component_interface_utils/rclcpp/topic_subscription.hpp [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef COMPONENT_INTERFACE_UTILS__RCLCPP__TOPIC_SUBSCRIPTION_HPP_
+#define COMPONENT_INTERFACE_UTILS__RCLCPP__TOPIC_SUBSCRIPTION_HPP_
+
+#include <rclcpp/subscription.hpp>
+
+namespace component_interface_utils
+{
+
+/// The wrapper class of rclcpp::Subscription.
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/topic_subscription.hpp:23]: Same as https://github.com/autowarefoundation/autoware.universe/pull/899#discussion_r873347038

File: common/component_interface_utils/include/component_interface_utils/response.hpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#ifndef COMPONENT_INTERFACE_UTILS__RESPONSE_HPP_
+#define COMPONENT_INTERFACE_UTILS__RESPONSE_HPP_
+
+#include <autoware_ad_api_msgs/msg/response_status.hpp>
+#include <autoware_ad_api_msgs/msg/response_status_detail.hpp>
+
+namespace component_interface_utils::response
+{
+
+using ResponseStatusDetail = autoware_ad_api_msgs::msg::ResponseStatusDetail;
```
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/response.hpp]: Removed https://github.com/autowarefoundation/autoware.universe/pull/899/commits/e1a602fa353845176270d4759f0a58fb3748cc09

File: system/default_ad_api/src/interface.cpp (hunk 1 above)
Comment (isamu-takagi) [reply system/default_ad_api/src/interface.cpp:15]: Yes. That is preferable. However, this refactoring depends on this PR. It can't be separated because the build fails.

File: common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kenji-miyake) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:29]: So do you mean if we use inheritance, the log output won't work even if we add it, right? :thinking:

[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:29]: if using inheritance, we need to override all functions. For example, if rclcpp changes and add new member function, that function will not output a log. if not using inheritance, we can detect the use of that function by build errors until logging version added.

[the remark this answers is not in the document]
Comment (kenji-miyake) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:29]: I understand. Thank you for your explanation.
How about writing it in the README or Doxygen comments?

[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/service_client.hpp:29]: Add design section to README https://github.com/autowarefoundation/autoware.universe/pull/899/commits/48867729a6fcdf7c3c4c773ea6982c9161f15a3d

File: system/default_ad_api/src/interface.cpp (hunk 1 above)
Comment (kenji-miyake) [reply system/default_ad_api/src/interface.cpp:15]: Okay, I understand!

File: common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kenji-miyake) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp:23]: @isamu-takagi I feel it's better to keep that background as comments. I'll leave it to you whether you write the comments.

**Review by kenji-miyake — APPROVED** [review]

File: common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (isamu-takagi) [reply common/component_interface_utils/include/component_interface_utils/rclcpp/topic_publisher.hpp]: Add comment https://github.com/autowarefoundation/autoware.universe/pull/899/commits/c34ca65b2d0c4522a16fbff3d30a91ce73491b58
